### PR TITLE
Initial implementation of messenger 

### PR DESCRIPTION
**Pull request by Wondertan**
v0 implementation for messenger. It still requires more documentation and testing, along with broadening of the design document, but that's a bare minimum required to migrate at least one Tendermint protocol.

To review the API looking at `messenger.go` is enough. For a more in-depth business logic review, one should look into `messenger_in/out.go`

### CODE REVIEW

**Comment by liamsi** [comment]
This is a bit clearer than #1. 

_Before_ continuing to make the code nice and tested I'd recommend going one of the following routes first: 
- either communicate the design and trajectory of this library in the context of tendermint in a document, or,
- showcase in a draft PR how usage in tendermint could look like.

**Comment by Wondertan** [comment]
As discussed in sync, I am going to present the messenger right after at least one Tendermint reactor is migrated.

File: messenger_in.go
Comment (liamsi) [inline messenger_in.go:69]: Do you have to use reflection here?

**Review by liamsi — COMMENTED** [review]
Again the code looks great. As this is growing I think we really need to communicate what the purpose of this library will be. 
I'm wondering if we shouldn't just narrow this down to use protobuf instead of our [own interface](https://github.com/celestiaorg/go-libp2p-messenger/blob/f4ba7396b9dde34ac24656556b140bfdb7f6a573/serde/serde.go#L11)?

Orthogonal to this PR: I'd also recommend renaming the Marshal functions that call the actual marshal method of the message to sth that communicates the length prefixing as this makes it more explicit (note that most protobuf libs provide helpers for this already).

Q: Any chance we can get rid of the use of reflection even if that means we'd lose some generality here?

**Comment by liamsi** [comment]
Here is my current understanding of things. Please correct me if I'm misunderstanding anything. The goal of this library is to replace the router in [this design](https://github.com/tendermint/tendermint/blob/master/docs/architecture/adr-062-p2p-architecture.md#detailed-design):


Essentially the diagram stays the same besides that the router would disappear and replace with the Messenger here? All the lower-level abstractions like connections, transports, streams, multiplexing would be replaced with the libp2p equivalents. I'm not completely sure who would do the mapping between channels and reactors then (would the router need to stay but be simpler?).

Reactors and their channels stay as they are and I assume the PeerManagement can also stay in tendermint where it makes sense? A diagram and a simple example like [this one](https://github.com/tendermint/tendermint/blob/master/docs/architecture/adr-062-p2p-architecture.md#reactor-example) would be helpful in actually reviewing this a lot. I can create that diagram for you if my understanding above is correct and for the example, we can still aim for the actual tendermint reactor example to save us some time.

**Comment by Wondertan** [comment]
> I'm wondering if we shouldn't just narrow this down to use `protobuf` 

@liamsi, Nope! One of the design goals for messenger is not to dictate the wire protocol. Users should be able to choose any serialization they want. Also, the messenger does not send any data on the wire. If we use `protobuf` by default, we would wrap users' payload with our custom message, making the messenger a subprotocol in a stack, but instead, we should try to avoid this and make messenger more like a service, rather than a utility protocol in a stack.

Also, the Message interface does not know anything about serialization, though it dictates its form slightly through methods. But those methods are inspired by `protobuf` in the first place, so we support `protobuf` by default. Typical `protobuf` messages generate by `gogo/proto` are similarly defined with `Size|MarshalTo|Unmarshal` methods.

> I'd also recommend renaming the Marshal functions that call the actual marshal method of the message to sth that communicates the length prefixing as this makes it more explicit (note that most protobuf libs provide helpers for this already).

The `MarshalTo` method in structures generated by `gogo/proto` is not length-prefixed, but the `Marshal` is and we don't use it, so I don't get the recommendation here.

> Q: Any chance we can get rid of the use of reflection even if that means we'd lose some generality here?

I tried several things, but generality forces us to ease it, but let's dive into why it's not an issue. The common concern about reflection is performance and with the current approach, we avoid slowness! The only place where reflection is used - is in an hotpath where you pointed in the review. This place allocates the arbitrary type, similar to `new` from an already know type stored in the Messenger. Usually, the code that [figures out the type](https://github.com/celestiaorg/go-libp2p-messenger/blob/messenger/options.go#L22) is slow, not the one that does the allocations and [benchmarks](https://gist.github.com/nfisher/a9aa2d34fedb338667bf) show that. It is still slower than regular `new`, but is not that bad. 

Another argument here is that even if we consider the allocations are really slow with the approach, we are still good to go, as the code which goes after the allocation is IO-bound, while allocations are CPU-bound and in comparison, the time it would take for an allocation is nothing against the time it takes to wait for the actual message.

Moreover, once generics land we can easily swap out the reflection here with something like `Messenger[YourMsgType]`. This is a thing that makes messenger also future-proof.

**Comment by liamsi** [comment]
> One of the design goals for messenger is not to dictate the wire protocol.

It's unclear why this is a design goal or why it is important. It seems like optimizing for a general-purpose scenario while the only description in the Readme speaks of "an effort to migrate Tendermint over libp2p" as the usage scenario. Both ecosystems heavily make use of protobuf. It does seem like the goal is kinda set arbitrarily and is opinionated and not really driven by the use-case communicated in the readme? While I think it's great to be able to swap out the serialization format from an aesthetic and an API design perspective, I don't follow how this makes us reach the actual goal ("migrate Tendermint over libp2p").

> The MarshalTo method in structures generated by gogo/proto is not length-prefixed, but the Marshal is and we don't use it, so I don't get the recommendation here.

That is why I'm suggesting renaming the MarshalTo method in this repo (e.g. to MarshalLenPrefixed). Otherwise, it is rather very implicit what is going on.

But let's not argue about these details! We can iterate over this another time. All this is not that important but points to a deeper issue here:

I think what is much more important is to get the repo actually communicate what its goals are and where in above's diagram  the messenger would fit and what its tasks actually are! I think that should be the priority (besides trying to actually have an example reactor using the messenger). We are kinda navigating in the void. I'm can approve this PR but the trajectory is really non-obvious (to anyone besides you).
(see: https://github.com/celestiaorg/go-libp2p-messenger/pull/2#issuecomment-924104795)

**Review by liamsi — APPROVED** [review]

**Pull request merged**
